Pyrado/pyrado/policies/recurrent/two_headed_rnn.py

```diff
@@ -176,25 +176,65 @@ def evaluate(self, rollout: StepSequence, hidden_states_name: str = "hidden_stat
         # Set policy, i.e. PyTorch nn.Module, to evaluation mode
         self.eval()
 
-        act_list = []
-        head2_list = []
+        # Get a list of all observation sequences and their respective lengths
+        obs_list = []
+        lengths = []
+        hidden_list = []
+
         for ro in rollout.iterate_rollouts():
             if hidden_states_name in rollout.data_names:
-                # Get initial hidden state from first step
-                hidden = ro[0][hidden_states_name]
+                # Get initial hidden state from first step, but do not unpack it yet
+                hidden_list.append(ro[0][hidden_states_name])
             else:
-                # Let the network pick the default hidden state
-                hidden = None
-            # Run steps consecutively reusing the hidden state
-            for step in ro:
-                act, head2, hidden = self(step.observation, hidden)
-                act_list.append(act)
-                head2_list.append(head2)
+                # If no hidden state is given, let it be None
+                hidden_list.append(None)
+
+            # Reshape observations to match PyTorch's RNN sequence protocol
+            obs = ro.get_data_values("observations", True)
+            obs = obs.to(device=self.device)
+
+            # Get all observation sequences and their respective lengths
+            obs_list.append(obs)
+            lengths.append(len(obs))
+
+        # Pad and then pack observations
+        obs_padded = to.nn.utils.rnn.pad_sequence(obs_list)
+        obs_packed = to.nn.utils.rnn.pack_padded_sequence(obs_padded, lengths=lengths, enforce_sorted=False)
```

Comment:
I know you did it the same way for the "casual" RNNs, but why are we not doing `enforce_sorted=True`?

Reply:
from the PyTorch wiki: _enforce_sorted (bool, optional) – if True, the input is expected to contain sequences sorted by length in a decreasing order. If False, the input will get sorted unconditionally_

I don't think that we get the observations sequences in a decreasing order since the rollouts themselves are not sorted in this way. Since they are unsorted, using this option, PyTorch will sort them for us.

Reply:
thank you for the clarification :+1: